src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java

```diff
@@ -1,647 +1,560 @@
-    {
-        return (BuildingFarmer) (worker.getWorkBuilding());
-    }
+    /**
+     * Changed after finished harvesting in order to dump the inventory.
+     */
+    private boolean shouldDumpInventory = false;
 
     /**
-     *
+     * The offset to work at relative to the scarecrow.
      */
-    private void checkForCrops()
-    {
-        if (crops.isEmpty() && !crops2.isEmpty())
-        {
-            crops.addAll(crops2);
-            crops2.clear();
-            job.setStage(Stage.NEED_SEEDS);
-        }
-    }
+    private BlockPos workingOffset;
 
-    @Override
-    protected void updateRenderMetaData()
+    /**
+     * The standard delay the farmer should have.
+     */
+    private static final int STANDARD_DELAY = 5;
+
+    /**
+     * The bonus the farmer gains each update is level/divider.
+     */
+    private static final int DELAY_DIVIDER = 10;
+
+    /**
+     * The delay the farmer should have each action: hoeing, planting, harvesting.
+     */
+    private int workingDelay = STANDARD_DELAY - this.worker.getLevel() / DELAY_DIVIDER;
+
+    /**
+     * Defines if the farmer should request seeds for the current field.
+     */
+    private boolean requestSeeds = true;
+
+    /**
+     * Defines if the farmer should try to get the seeds from his chest.
+     */
+    private boolean shouldTryToGetSeed = true;
+
+    /**
+     * Constructor for the Farmer.
+     * Defines the tasks the Farmer executes.
+     *
+     * @param job a farmer job to use.
+     */
+    public EntityAIWorkFarmer(JobFarmer job)
     {
-        String renderMetaData = getRenderMetaSeeds();
-        //TODO: Merge this into worker
-        worker.setRenderMetadata(renderMetaData);
+        super(job);
+        super.registerTargets(
+                new AITarget(IDLE, () -> START_WORKING),
+                new AITarget(START_WORKING, this::startWorkingAtOwnBuilding),
+                new AITarget(PREPARING, this::prepareForFarming),
+                new AITarget(FARMER_HOE, this::hoe),
+                new AITarget(FARMER_PLANT, this::plant),
+                new AITarget(FARMER_HARVEST, this::harvest)
+        );
+        worker.setSkillModifier(2*worker.getCitizenData().getEndurance() + worker.getCitizenData().getCharisma());
     }
 
     /**
-     * @return String for seeds renderer if farmer has seeds, otherwise empty string
+     * Redirects the fisherman to his building.
+     * @return the next state.
      */
-    private String getRenderMetaSeeds()
+    private AIState startWorkingAtOwnBuilding()
     {
-        if (hasSeed())
+        if (walkToBuilding())
         {
-            return RENDER_META_SEEDS;
+            return getState();
         }
-        return "";
+        return PREPARING;
     }
 
-    private boolean hasSeed()
+    /**
+     * Prepares the farmer for farming and
+     * requests the tools and checks if the farmer has sufficient fields.
+     * @return the next AIState
+     */
+    private AIState prepareForFarming()
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of returns of this method 7, down to the maximum allowed 5. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "prepareForFarming" is 22 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)